libs/ktem/ktem/reasoning/prompt_optimization/fewshot_rewrite_question.py

```diff
@@ -0,0 +1,115 @@
+import langchain_core.messages
+from ktem.llms.manager import llms
+from ktem.reasoning.prompt_optimization.rewrite_question import (
+    DEFAULT_REWRITE_PROMPT,
+    RewriteQuestionPipeline,
+)
+from langchain.prompts import (
+    ChatPromptTemplate,
+    FewShotChatMessagePromptTemplate,
+    SemanticSimilarityExampleSelector,
+)
+from langchain_community.vectorstores import Chroma
+from langchain_openai import OpenAIEmbeddings
+from theflow.settings import settings as flowsettings
+
+from kotaemon.base import AIMessage, Document, HumanMessage, Node, SystemMessage
+from kotaemon.llms import ChatLLM
+
+
+class FewshotRewriteQuestionPipeline(RewriteQuestionPipeline):
+    """Rewrite user question
+
+    Args:
+        llm: the language model to rewrite question
+        rewrite_template: the prompt template for llm to paraphrase a text input
+        lang: the language of the answer. Currently support English and Japanese
+    """
+
+    llm: ChatLLM = Node(default_callback=lambda _: llms.get_default())
+    rewrite_template: str = DEFAULT_REWRITE_PROMPT
+    lang: str = "English"
+    final_prompt: ChatPromptTemplate
+
+    def create_example_selector(
+        self, examples, k: int = getattr(flowsettings, "N_PROMPT_OPT_EXAMPLES", 0)
+    ):
+        if not examples:
+            return None
+
+        embeddings = OpenAIEmbeddings(
+            openai_api_key=getattr(flowsettings, "KH_EMBEDDINGS")["openai"]["spec"][
+                "api_key"
+            ],
+            openai_api_type="openai",
+            openai_api_base=getattr(flowsettings, "KH_EMBEDDINGS")["openai"]["spec"][
+                "base_url"
+            ],
+            openai_api_version=getattr(flowsettings, "KH_EMBEDDINGS")["openai"]["spec"][
+                "api_version"
+            ],
+        )
+        vectorstore = Chroma(
+            "rephrase_question",
+            embeddings,
+            persist_directory=str(getattr(flowsettings, "KH_APP_DATA_DIR")),
+        )
```

Comment:
OpenAI models might not be available during runtime. 

A pipeline/component should work with any embedding model and any vector store regardless of their classes. This should use `kotaemon`'s abstract classes for embedding model and vector store instead of `OpenAIEmbeddings` and `Chroma` explicitly. It is also recommended to use `kotaemon`'s classes instead of 3rd party ones to avoid reliance on dependencies.

Moreover, vector stores and embedding models are configured and managed globally. For how to access them, please check out other pipelines' implementations or consult @trducng.